techlibs/quicklogic/synth_quicklogic.cc

```diff
@@ -0,0 +1,141 @@
+#include "kernel/register.h"
+#include "kernel/celltypes.h"
+#include "kernel/rtlil.h"
+#include "kernel/log.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+struct SynthQuickLogicPass : public ScriptPass
+{
+    SynthQuickLogicPass() : ScriptPass("synth_quicklogic", "Synthesis for QuickLogic FPGAs") { }
+
+    void help() override
+    {
+        log("\n");
+        log("   synth_quicklogic [options]\n");
+        log("This command runs synthesis for QuickLogic FPGAs\n");
+        log("\n");
+        log("    -top <module>\n");
+        log("         use the specified module as top module\n");
+        log("\n");
+        log("    -edif <file>\n");
+        log("        write the design to the specified edif file. writing of an output file\n");
+        log("        is omitted if this parameter is not specified.\n");
+        log("\n");
+        log("    -blif <file>\n");
+        log("        write the design to the specified BLIF file. writing of an output file\n");
+        log("        is omitted if this parameter is not specified.\n");
+        log("\n");
+        help_script();
+        log("\n");
+    }
+
+    std::string top_opt = "";
+    std::string edif_file = "";
+    std::string blif_file = "";
+    std::string currmodule = "";
+
+    void clear_flags() override
+    {
+        top_opt = "-auto-top";
+        edif_file.clear();
+        blif_file.clear();
+    }
+
+    void execute(std::vector<std::string> args, RTLIL::Design *design) override
+    {
+        std::string run_from, run_to;
+        clear_flags();
+
+        size_t argidx;
+        for (argidx = 1; argidx < args.size(); argidx++)
+        {
+            if (args[argidx] == "-top" && argidx+1 < args.size()) {
+                top_opt = "-top " + args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-edif" && argidx+1 < args.size()) {
+                edif_file = args[++argidx];
+                continue;
+            }
+            if (args[argidx] == "-blif" && argidx+1 < args.size()) {
+                blif_file = args[++argidx];
+                continue;
+            }
+            break;
+        }
+        extra_args(args, argidx, design);
+
+        if (!design->full_selection())
+            log_cmd_error("This command only operates on fully selected designs!\n");
+
+        log_header(design, "Executing SYNTH_QUICKLOGIC pass.\n");
+        log_push();
+
+        run_script(design, run_from, run_to);
+
+        log_pop();
+    }
+
+    void script() override
+    {
+        if (check_label("begin")) {
+            run("read_verilog -lib +/quicklogic/cells_sim.v");
+            run(stringf("hierarchy -check %s", top_opt.c_str()));
+        }
+
+        if (check_label("prepare")) {
```

Comment:
The flow structure here seems quite weird and mismatched with other synth flows.  While yosys isn't very good at consistency here (something we should fix one day...), there are some major issues with this pass set:

- `memory` is never called, so Verilog/RTLIL memories will just end up at the output stage and break it
- `fsm` is never called
- `cmp2lcu` is not used (though that may not be a problem since you don't seem to have a `$lcu` model?)
- LUTs and FFs are mapped in the same step. While this works, it is suboptimal — for one, it can prevent merging FF mapping detritus (like the reset/set/enable inverters) with surrounding logic.  The standard approach in newer targets is to map FFs in a separate step before abc and LUT mapping.

These three steps (prepare/coarse/map) should be replaced by something resembling `synth_intel_alm`'s flatten/coarse/map_ffram/map_ffs/map_luts steps — it's probably our cleanest pass at the moment (it seems the only thing you will need to add wrt intel_alm is `muxpack` + `muxcover`, the rest is already there).

Reply:
Also, an update on the current FF type refactoring effort: recently, I added the `dfflegalize` pass which converts all yosys FF cells to ones actually supported by the target, and will soon be required in all synth flows.  This will actually simplify your mapping code, since now you're only required to describe the exact "native" cell types and don't have to eg. manually add inverters to match control signal polarity, or upconvert to more capable cell types — just list the cell types you directly support on the `dfflegalize` command line and only keep techmap rules for those.